### PR TITLE
Updating AppleCider to use current Hyrax

### PR DESCRIPTION
**Pull request by drewoldag**
# Updating AppleCider to use the current version of Hyrax (v0.8.3)
- Updated dataset classes to import `dataset` instead of `data_set`
- Updated models
  - Renamed `train_step` to `train_batch`
  - Added `infer_batch` (just calls `forward`)
  - Renamed `to_tensor` to `prepare_inputs`
- Updated smoke tests, unit tests, and pyproject.toml to specify python >= 3.11
- Updated `model_input`s to `data_request`s in the example notebook and configs
- Made sure that example notebooks can run (just a smoke test)
- Verified `to_onnx` and `engine` still work
- Removed dataset-level `collate` from photo_dataset, replaced with `collate_photometry`.
- Double check the returned values from photo_dataset methods `get_mean` and `get_std`. I had to reshape them to get things to work.

## TODO
- Wait for Felipe to review the padding. 
- ~~Add `validate_batch` methods in astrominn and spectranet.~~ Won't do this, can revisit if we break apart this repo into distinct models.

### CODE REVIEW

File: src/applecider/models/spectranet.py [hunk 1]
```diff
@@ -169,7 +169,7 @@ def forward(self, batch):
         else:
             return self.classifier(x_fused)
 
-    def train_step(self, batch):
+    def train_batch(self, batch):
```
[the remark this answers is not in the document]
Comment (drewoldag) [reply src/applecider/models/spectranet.py:172]: Fixed description on this PR

File: src/applecider/datasets/image_and_metadata_dataset.py [hunk 2]
```diff
@@ -59,13 +57,13 @@ def __init__(self, config, data_location):
         super().__init__(config)
         # Additional initialization for image and metadata dataset can be added here
 
-    def get_metadata(self, index):
+    def get_metadata(self, index) -> np.ndarray:
         # Method to retrieve metadata at the specified index
         if self.use_oversampling:
             index, is_oversampled = self.retrieve_oversampled_index(index)
-        return self.raw_files[index].get("metadata")
+        return self.raw_files[index].get("metadata").numpy()
```
[the remark this answers is not in the document]
Comment (drewoldag) [reply src/applecider/datasets/image_and_metadata_dataset.py]: I think that we can get away with just np.asarray(...). Fixed as such.

File: src/applecider/datasets/photo_dataset.py [hunk 3]
```diff
@@ -42,21 +42,13 @@ def __init__(self, config: dict, data_location: Union[Path, str] = None, horizon
...
         # Find the row in the manifest ids
         old_idx = idx
         if self.use_oversampling:
             idx, is_oversampled = self.retrieve_oversampled_index(idx)
-        return self.object_ids[idx]
-
-    def ids(self):
-        for idx in range(len(self)):
-            yield self.get_object_id(idx)
+        return str(self.object_ids[idx])
```
[the remark this answers is not in the document]
Comment (drewoldag) [reply src/applecider/datasets/photo_dataset.py:50]: Removed old_idx from get_object_id and get_label.

File: src/applecider/models/HyraxBaselineCLS.py [hunk 4]
```diff
@@ -283,6 +286,59 @@ def train_step(self, batch):
...
+        loss_b = F.cross_entropy(b_hat.view(-1, 3)[mf], true_b[mf])
+        dt_gt = torch.roll(data[..., 1], -1, dims=1)
+        dt_gt[:, -1] = 0.0
+        dt_gt = dt_gt.view(-1)
+        loss_dt = F.mse_loss(dt_hat[..., 0].view(-1)[mf], dt_gt[mf])
+
+        lambda_f = self.config["model"]["HyraxBaselineCLS"]["lambda_f"]
+        lambda_b = self.config["model"]["HyraxBaselineCLS"]["lambda_b"]
+        lambda_dt = self.config["model"]["HyraxBaselineCLS"]["lambda_dt"]
+        loss = lambda_f * loss_f * lambda_b * loss_b * lambda_dt * loss_dt
```
[the remark this answers is not in the document]
Comment (drewoldag) [reply src/applecider/models/HyraxBaselineCLS.py]: This seems like the right choice to make. I've made the update.

File: src/applecider/models/HyraxBaselineCLS.py [hunk 5]
```diff
@@ -119,8 +119,11 @@ def train_step(self, batch):
         # accuracy, total loss/tot n, any custom metrics
         return {"loss": loss.item(), "num_tdes": np.sum([labels.cpu().numpy() == 4])}
 
+    def infer_batch(self, batch):
+        return self.forward(batch)
+
     @staticmethod
-    def to_tensor(data_dict):
+    def prepare_inputs(data_dict):
```
[the remark this answers is not in the document]
Comment (drewoldag) [reply src/applecider/models/HyraxBaselineCLS.py:126]: I'll need someone with more awareness of this model to double check on this.

File: src/applecider/models/HyraxBaselineCLS.py [hunk 6]
```diff
@@ -275,14 +278,67 @@ def train_step(self, batch):
         lambda_f = self.config["model"]["HyraxBaselineCLS"]["lambda_f"]
         lambda_b = self.config["model"]["HyraxBaselineCLS"]["lambda_b"]
         lambda_dt = self.config["model"]["HyraxBaselineCLS"]["lambda_dt"]
-        loss = lambda_f * loss_f * lambda_b * loss_b * lambda_dt * loss_dt
+        loss = lambda_f * loss_f + lambda_b * loss_b + lambda_dt * loss_dt
```
Comment (drewoldag) [inline src/applecider/models/HyraxBaselineCLS.py:281]: Double check that this is correct.

File: src/applecider/models/HyraxBaselineCLS.py [hunk 7]
```diff
@@ -275,14 +278,67 @@ def train_step(self, batch):
...
+        loss_b = F.cross_entropy(b_hat.view(-1, 3)[mf], true_b[mf])
+        dt_gt = torch.roll(data[..., 1], -1, dims=1)
+        dt_gt[:, -1] = 0.0
+        dt_gt = dt_gt.view(-1)
+        loss_dt = F.mse_loss(dt_hat[..., 0].view(-1)[mf], dt_gt[mf])
+
+        lambda_f = self.config["model"]["HyraxBaselineCLS"]["lambda_f"]
+        lambda_b = self.config["model"]["HyraxBaselineCLS"]["lambda_b"]
+        lambda_dt = self.config["model"]["HyraxBaselineCLS"]["lambda_dt"]
+        loss = lambda_f * loss_f + lambda_b * loss_b + lambda_dt * loss_dt
```
Comment (drewoldag) [inline src/applecider/models/HyraxBaselineCLS.py:335]: Double check that this is correct.

File: src/applecider/models/HyraxBaselineCLS.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (drewoldag) [reply src/applecider/models/HyraxBaselineCLS.py:126]: @FeradoFogo if you have a moment, would you mind taking a look at this?

**Review by asasli — APPROVED** [review]
Everything looks fine. Just check with @FeradoFogo the padding.

**Review by FeradoFogo — CHANGES_REQUESTED** [review]
This looks correct to me relative to the original implementation. The collate step still leaves the photometry input capped at `max_len` (257 for us), and `forward` is still where the CLS token is prepended and the padding mask gets extended. The one thing I would change is the fallback case when `pad_mask` is missing: right now `prepare_inputs` builds it with length `L + 1`, but then forward adds the CLS mask again, so I think it should start as `(B, L)` instead of `(B, L + 1)`.

For a future PR, we could optimize the `collate_photometry` memory allocation for long light curves by truncating first and padding directly to `max_len`, instead of padding to the longest sequence and slicing back.

File: src/applecider/models/HyraxBaselineCLS.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (drewoldag) [reply src/applecider/models/HyraxBaselineCLS.py:126]: Removed the `+1` from the length of `false_mask`, so we shoule be good to go now. @FeradoFogo if you could confirm, that would be great :)

**Review by FeradoFogo — APPROVED** [review]
